src/traces/box/hover.js

```diff
@@ -14,94 +14,167 @@ var Fx = require('../../components/fx');
-        pointData2[valLetter + '0'] = pointData2[valLetter + '1'] = val;
-        pointData2[valLetter + 'LabelVal'] = di[attr];
-        pointData2.attr = attr;
+                // copy out to a new object for each value to label
+                var val = valAxis.c2p(di[attr], true);
+                pointData2 = Lib.extendFlat({}, pointData);
+                pointData2[valLetter + '0'] = pointData2[valLetter + '1'] = val;
+                pointData2[valLetter + 'LabelVal'] = di[attr];
+                pointData2.attr = attr;
 
-        if(attr === 'mean' && ('sd' in di) && trace.boxmean === 'sd') {
-            pointData2[valLetter + 'err'] = di.sd;
+                if(attr === 'mean' && ('sd' in di) && trace.boxmean === 'sd') {
+                    pointData2[valLetter + 'err'] = di.sd;
+                }
+                // only keep name on the first item (median)
+                pointData.name = '';
+
+                closeData.push(pointData2);
+            }
         }
-        pointData.name = ''; // only keep name on the first item (median)
-        closeData.push(pointData2);
     }
+
+    if(hoveron.indexOf('points') !== -1) {
+        var xPx = xa.c2p(xval);
+        var yPx = ya.c2p(yval);
+
+        // do not take jitter into consideration in compare hover modes
+        var kx, ky;
+        if(hovermode === 'closest') {
+            kx = 'x';
+            ky = 'y';
+        } else {
+            kx = 'xh';
+            ky = 'yh';
+        }
+
+        dx = function(di) {
+            var rad = Math.max(3, di.mrc || 0);
+            return Math.max(Math.abs(xa.c2p(di[kx]) - xPx) - rad, 1 - 3 / rad);
+        };
+        dy = function(di) {
+            var rad = Math.max(3, di.mrc || 0);
+            return Math.max(Math.abs(ya.c2p(di[ky]) - yPx) - rad, 1 - 3 / rad);
+        };
+        distfn = Fx.getDistanceFunction(hovermode, dx, dy);
+
+        for(i = 0; i < cd.length; i++) {
+            di = cd[i];
+
+            for(j = 0; j < (di.pts || []).length; j++) {
+                var pt = di.pts[j];
+
+                var newDistance = distfn(pt);
+                if(newDistance <= pointData.distance) {
+                    pointData.distance = newDistance;
+
+                    var xc = xa.c2p(pt.x, true);
+                    var yc = ya.c2p(pt.y, true);
+                    var rad = pt.mrc || 1;
+
+                    pointData2 = Lib.extendFlat({}, pointData, {
+                        // corresponds to index in x/y input data array
+                        index: pt.i,
+                        color: marker.color,
+                        x0: xc - rad,
+                        x1: xc + rad,
+                        xLabelVal: pt.x,
+                        y0: yc - rad,
+                        y1: yc + rad,
+                        yLabelVal: pt.y
+                    });
+
+                    closeData.push(pointData2);
+                }
+            }
+        }
+    }
+
     return closeData;
```

Comment:
Another debatable choice: this commit allows `'boxes'` and `'points'` hoveron picks to coexist. With `hoveron: 'boxes+points'`, we get:

![peek 2017-10-16 23-51](https://user-images.githubusercontent.com/6675409/31645973-f1394bee-b2cc-11e7-9d6b-04b8c4b87bce.gif)

where box and point hover label can be displayed at the same time. Does anyone find this potentially problematic?

Reply:
... and under `hovermode: 'x'`, this can happen:

![peek 2017-10-17 00-05](https://user-images.githubusercontent.com/6675409/31646272-e6398090-b2ce-11e7-9ca3-b2f8ec75ea91.gif)

Reply:
`boxes+points` could be useful, but I think we want to show fewer points. Currently it seems to return all the points that are in range, but most real-world box uses have far too many points to label them all, and we should match `scatter` which chooses the *best* point when more than one is in range.

So I guess I'm proposing:
- In `closest` mode, show only one point or stats for one box, and points have priority
  - If there's a point in range and `hoveron` has `points`, show the best single point only.
  - If `hoveron` has `boxes` and there's no point in range (or `hoveron` doesn't have `points`), show the box stats.
- In `compare` mode I don't really see any extra useful thing to do with points - I think we should still only show one point per trace, so I guess that means not considering only x when choosing the best point? I know that's not consistent with other types, but I feel like usability is much more important.
  - Perhaps though if `hoveron = 'boxes+points'` we should allow a point AND the box stats to be labeled?
  - Already if there are multiple boxes in range (ie `boxmode = 'overlay'`) we'll see stats for all of them, right? I think that's good.

One other thought: do we want to do something to distinguish statistics labels from point labels? Perhaps actually say what statistical measure is being shown? Which might be nice on its own merits! But we'd have to be careful about the whisker ends, whether they represent upper/lower fence or max/min.

Reply:
> > but most real-world box uses have far too many points to label them all, and we should match scatter

Good point here. I'll implement :arrow_heading_up:

Reply:
Here's what I come up with: https://github.com/plotly/plotly.js/pull/2094/commits/f228a5ec6e232d6c1427fdef37b77a252e018eff

> One other thought: do we want to do something to distinguish statistics labels from point labels? Perhaps actually say what statistical measure is being shown?

The above commit does not implement this. I think it's a good idea though. Should the statistical measures always show up next to their value in the hover labels or just when point and box hover label show simultaneously?

Reply:
> Should the statistical measures always show up next to their value in the hover labels or just when point and box hover label show simultaneously?

I think it would be nice to always show them, if it doesn't look too cluttered.

Reply:
Here's a quick-and-easy version:

![peek 2017-10-19 12-02](https://user-images.githubusercontent.com/6675409/31781007-5605305c-b4c5-11e7-816e-f3d78b5a79ca.gif)

we might have to expand out `uf` and `lf`. Spelling out _upper fence_ and _lower fence_ looks like:

![peek 2017-10-19 12-06](https://user-images.githubusercontent.com/6675409/31781246-00529716-b4c6-11e7-9c7b-eab05068fcef.gif)

does that look too cluttered to you?

Reply:
👍  for spelled out, I think you're right that `uf` / `lf` will confuse people.

Reply:
maybe even spell out `median`

Reply:
full version:

![peek 2017-10-19 13-04](https://user-images.githubusercontent.com/6675409/31783959-0d5a0cb6-b4ce-11e7-8e50-e959d4261183.gif)

Reply:
done in a3ec75a (with :hammer: :palm_tree: commit b66628e)